File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.scala
object Solution {
    def kthSmallest(root: TreeNode, k: Int): Int = {
        val (_, _, res) = helper(root, k)
        res
    }

    def helper(root: TreeNode, k: Int): (Boolean, Int, Int) = {
        if (root == null) (false, 0, 0)
        else {
            val (lf, ls, lr) = helper(root.left, k)
            if (lf) (lf, ls, lr) 
            else {
                if (ls == k - 1) (true, k, root.value)
                else {
                    val (rf, rs, rr) = helper(root.right, k - ls - 1)
                    if (rf) (rf, rs, rr)
                    else (false, ls + rs + 1, 0)
                }
            }
        }
    }
}

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode result = null;
    
    public int kthSmallest(TreeNode root, int k) {
        helper(root, 0, k);
        return result.val;
    }
    
    int[] helper(TreeNode root, int current, int k) {
        if (root == null) return new int[] { current, -1 };
        int[] res = helper(root.left, current, k);
        if (res[1] == 1) return new int[] { -1, 1 };
        current = res[0] + 1;
        if (current == k) {
            result = root;
            return new int[] {-1, 1};
        }
        return helper(root.right, current, k);
    }
}